core-application/src/test/java/gohigher/usecase/OAuth2CommandServiceTest.java

```diff
@@ -0,0 +1,91 @@
+package gohigher.usecase;
+
+import static org.assertj.core.api.Assertions.*;
+import static org.junit.jupiter.api.Assertions.*;
+import static org.mockito.BDDMockito.*;
+
+import java.util.Optional;
+
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Nested;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.Mock;
+import org.mockito.junit.jupiter.MockitoExtension;
+
+import gohigher.port.out.UserPersistenceCommandPort;
+import gohigher.port.out.UserPersistenceQueryPort;
+import gohigher.user.Role;
+import gohigher.user.User;
+import gohigher.user.auth.Provider;
+
+@ExtendWith(MockitoExtension.class)
+@DisplayName("OAuth2CommandService 클래스의")
+class OAuth2CommandServiceTest {
+
+	@Mock
+	private UserPersistenceQueryPort userPersistenceQueryPort;
+
+	@Mock
+	private UserPersistenceCommandPort userPersistenceCommandPort;
+
+	private UserCommandService userCommandService;
+
+	@BeforeEach
+	void setUp() {
+		userCommandService = new UserCommandService(userPersistenceQueryPort, userPersistenceCommandPort);
+	}
+
+	@DisplayName("login 메서드는")
+	@Nested
+	class login {
+
+		@DisplayName("해당 이메일로 가입한 사용자가 있을 때")
+		@Nested
+		class exist {
+
+			@DisplayName("저장되어 있는 정보를 반환해야 한다.")
+			@Test
+			void success() {
+				// given
+				String email = "test@gmail.com";
+				User user = new User(email, Role.USER, Provider.GOOGLE);
+				given(userPersistenceQueryPort.findByEmail(email)).willReturn(Optional.of(user));
+
+				// when
+				User savedUser = userCommandService.login(email, Provider.GOOGLE);
+
+				// then
+				assertThat(savedUser).isEqualTo(user);
+			}
+		}
```

Comment:
nested로 `~~할 때`라는 클래스의 displayName에서 주는 의미가 데이터가 주어진 것을 의미하는데 given의 부분을 상위로 올리는 것은 어떨까요?

```suggestion
		@DisplayName("해당 이메일로 가입한 사용자가 있을 때")
		@Nested
		class exist {
			// given
			String email = "test@gmail.com";
			User user = new User(email, Role.USER, Provider.GOOGLE);
				given(userPersistenceQueryPort.findByEmail(email)).willReturn(Optional.of(user));
			@DisplayName("저장되어 있는 정보를 반환해야 한다.")
			@Test
			void success() {
				// when
				User savedUser = userCommandService.login(email, Provider.GOOGLE);

				// then
				assertThat(savedUser).isEqualTo(user);
			}
		}
```

Reply:
저는 하나의 테스트 안에 테스트하고자 하는 내용이 명확히 드러나는 편이 좋다고 생각해요! 테스트하려는 기능 외에 셋팅해둬야 하는 경우가 있다면 `@BeforeEach` 등으로 미리 정의하는 것은 동의하나, when 을 실행시키기 위한 준비 과정을 분리하는건 한눈에 보기 어렵다고 생각합니다.

또한, 객체 생성 외에 메서드 호출이 given 영역에 포함되는 경우도 있는데, 이러한 경우 정의하기 번거로워 진다고 생각합니다.
(ex. find 를 호출하기 위해 save 등의 mocking 과정)

Reply:
저도 @BeforeEach로 지정하는 것을 생각했습니다..!
저는 개인적으로 계층형으로 클래스를 분리하게 된다면 테스트를 파악하는 범위를 테스트 메서드의 상위 클래스까지 봐야한다 생각합니다.
상위 클래스의 displayName만 보더라도 테스트를 하는 상황(given)을 나타내고 있다 생각합니다. 또한 어제 말한대로 여러 테스트를 할 때, 테스트 메서드뱔로 중복되는 given데이터가 계층형으로 구축하면 예방할 수 있다는 말들을 하였는데 이것은 상위 클래스에서 BeforeEach등을 통해 상위 클래스에서 given을 만들어주는 것으로부터 이점이 있다 생각합니다..!!
만약 계층형으로 나눈 클래스들이 단순히 DisplayName의 중복을 줄이는 역할만 한다면 계층의 depth를 줄이는 것이 좋다 생각합니다

Reply:
테스트 계층을 given/when/then 형식으로 구분했기 때문에 발생하는 고민인것 같습니다. 하지만 테스트의 구성이 어떻게 되더라도, '테스트 메서드 내에서 테스트 하고자 하는 것이 명확해야 한다' 라는 생각은 변하지 않습니다ㅠ 이는 값을 설정(given), 실행(when), 결과(then) 가 모두 한 곳에 포함되는 것을 의미합니다. 결국 어떤 값을 이용하여 실행했기 때문에 결과가 나타나는 것이기 때문입니다.

`@BeforeEach` 로 given 영역을 구분짓는 것은 역할이 다르다고 생각합니다. BeforeEach 로 설정하는 값은 보통 해당 범위 내 전역적으로 필요한 과정을 수행합니다. 복잡한 테스트를 작성하다보면 다른 테스트와 엮여, 동일한 과정을 수행해야 할 때 BeforeEach 를 사용합니다. 하지만 given 도 BeforeEach 로 처리할 경우, 이것이 when 에 입력하기 위해(given) 만들어진 값인지, when 을 실행하기 위해 전처리 되어야 하는 기능인지 구분하기 어려울것 같습니다.

Reply:
저는 사실 익숙하지 않아서인지 DCI패턴이 가독성에 그다지 좋은지는 모르겠어요. 저도 가독성을 위해서는 한 테스트 안에 given when then이 모두 포함되어 있어야 한다고 생각합니다. 실제로 우테코 미션에서 given 부분이 중복이 되더라도 그걸 밖으로 빼는 것은 가독성을 해칠 수 있다라는 피드백을 받고 공감했던 기억이 있습니다. 

하지만 성원님 생각과 마찬가지로 DCI 패턴을 사용하면서 코드는 테스트 하나하나에 다 들어있다면 그것 또한 큰 이익을 얻지 못한다고 생각합니다. 그래서 저는 DCI 패턴과 g,w,t 중 택해야한다면 후자가 더 좋은 것 같습니다!

( 물론 DCI 패턴이 뭔가 ci를 만들어서 displayname을 쫙볼때는 훨씬 편할 것 같긴해요)

Reply:
흠...저는 메서드 내의 `given/when/then`을 모두 적는 것에 대해 익숙해져 있기에 DCI 패턴에서 given을 외부로 빼는 것이 어색해보이는게 아닐까?생각합니다.

저는 개인적으로 DCI에서 Context부분을 Given으로 사용하지 않고 단순히 DisplayName용으로만 사용하는 것은 매번 테스트를 작성할 때마다 depth를 늘리며 inner class를 추가로 만드는 비용의 낭비라 생각합니다. 
가독성 측면에서도 하나의 TC를 살펴볼때 class depth를 2단계를 타고 들어와야하고 상황과 결과를 살펴볼 때 분리된 DisplayName을 봐야하여 가독성이 떨어진다 생각하고요..!

만약 DCI패턴이 어색하여 적용에 거부감이 드신다면 아래와 같이 method 단위로만 inner class를 사용하고 그 외의 상황과 결과에 대한 것은 하나의 메서드로 표현하는 것이 좋지 않을까 합니다.

```java
class ApplicationRepositoryTest {

	@DisplayName("findByUserIdAndMonth 메서드는")
	@Nested
	class Method_findByUserIdAndMonth {

		@DisplayName("여러 유저의 공고가 존재하여도 특정 유저의 데이터만 반환한다.")
		@Test
		void returns_only_data_for_a_specific_user() {
```

Reply:
저도 DisplayName만을 위해서 사용하는 거면 DCI가 과한 것 같습니다~!
개인적으로, DCI가 맘에 들었던 이유는 DisplayName이 아닌, `중복되는 given 절을 줄일 수 있었기 때문`입니다!

class depth를 타고 들어오거나, DisplayName 들을 보면서, 충분히 given 절을 파악할 수 있다고 생각합니다. given 절이 자신을 감싸고 있는 클래스에 정의되어 있지, 해당 테스트 파일 밖에 존재하는 것도 아니니까요! 또한, given 절이 10줄 이상과 같이 길어질 경우에는 given 절을 파악하기가 더 어렵다고 느꼈습니다!